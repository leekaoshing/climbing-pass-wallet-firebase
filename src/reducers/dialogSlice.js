import { createSlice } from '@reduxjs/toolkit';

export const dialogSlice = createSlice({
  name: 'dialog',
  initialState: {
    showAboutDialog: false,
    showAddGymDialog: false,
    showConfirmationDialog: false,
    showCreateUserDialog: false,
    showPasswordResetDialog: false,
    showUpdateResultDialog: false,
  },
  reducers: {
    setShowAboutDialog: (state, action) => {
      state.showAboutDialog = action.payload;
    },
    setShowAddGymDialog: (state, action) => {
      state.showAddGymDialog = action.payload;
    },
    setShowConfirmationDialog: (state, action) => {
      state.showConfirmationDialog = action.payload;
    },
    setShowCreateUserDialog: (state, action) => {
      state.showCreateUserDialog = action.payload;
    },
    setShowPasswordResetDialog: (state, action) => {
      state.showPasswordResetDialog = action.payload;
    },
    setShowUpdateResultDialog: (state, action) => {
      state.showUpdateResultDialog = action.payload;
    }
  }
});

export const {
  setShowAboutDialog,
  setShowAddGymDialog,
  setShowConfirmationDialog,
  setShowCreateUserDialog,
  setShowPasswordResetDialog,
  setShowUpdateResultDialog,
} = dialogSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectShowAboutDialog = state => {
  return state.dialog.showAboutDialog;
}

export const selectShowAddGymDialog = state => {
  return state.dialog.showAddGymDialog;
}

export const selectShowConfirmationDialog = state => {
  return state.dialog.showConfirmationDialog;
}

export const selectShowCreateUserDialog = state => {
  return state.dialog.showCreateUserDialog;
}

export const selectShowPasswordResetDialog = state => {
  return state.dialog.showPasswordResetDialog;
}

export const selectShowUpdateResultDialog = state => {
  return state.dialog.showUpdateResultDialog;
}

export default dialogSlice.reducer;
