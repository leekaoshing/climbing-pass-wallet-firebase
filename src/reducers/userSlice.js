import { createSlice } from '@reduxjs/toolkit';

export const userSlice = createSlice({
  name: 'user',
  initialState: {
    editableUser: null,
    loadingUpdateUser: false,
    updateResult: null,
  },
  reducers: {
    addGymToEditableUser: (state, action) => {
      state.editableUser.passes[action.payload] = 0
    },
    setEditableUser: (state, action) => {
      state.editableUser = action.payload;
    },
    setLoadingUpdateUser: (state, action) => {
      state.loadingUpdateUser = action.payload;
    },
    setUpdateResult: (state, action) => {
      state.updateResult = action.payload;
    }
  }
});

export const {
  addGymToEditableUser,
  setEditableUser,
  setLoadingUpdateUser,
  setUpdateResult
} = userSlice.actions;

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectEditableUser = state => {
  return state.user.editableUser;
}

export const selectLoadingUpdateUser = state => {
  return state.user.loadingUpdateUser;
}

export const selectUpdateResult = state => {
  return state.user.updateResult;
}

export const getPassDifferences = state => {
  const auth = state.firebase.auth;
  const oldPasses = state.firestore.data.users[auth.uid].passes;
  const newPasses = state.user.editableUser ? state.user.editableUser.passes : {};

  const differenceInPasses = {};
  Object.keys(newPasses).forEach(key => {
    const oldPassQuantity = oldPasses[key] === undefined ? 0 : oldPasses[key];
    const newPassQuantity = newPasses[key];
    const change = newPassQuantity - oldPassQuantity;
    if (change !== 0) differenceInPasses[key] = newPassQuantity - oldPassQuantity;
  })
  return differenceInPasses;
}

export default userSlice.reducer;
