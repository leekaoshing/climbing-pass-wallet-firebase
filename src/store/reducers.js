import { combineReducers } from 'redux'
import firebase from 'react-redux-firebase/lib/reducer'
import firestore from 'redux-firestore/lib/reducer'
import locationReducer from './reducers/location'
import dialogReducer from './reducers/dialog'
import userReducer from './reducers/user'

export function makeRootReducer(asyncReducers) {
  return combineReducers({
    // Add sync reducers here
    firebase,
    firestore,
    dialog: dialogReducer,
    user: userReducer,
    location: locationReducer,
    ...asyncReducers
  })
}

export function injectReducer(store, { key, reducer }) {
  store.asyncReducers[key] = reducer // eslint-disable-line no-param-reassign
  store.replaceReducer(makeRootReducer(store.asyncReducers))
}

export default makeRootReducer
