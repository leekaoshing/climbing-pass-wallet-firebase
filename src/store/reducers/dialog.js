import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  showUpdateResultDialog: false,
}

export const dialogSlice = createSlice({
  name: 'dialog',
  initialState,
  reducers: {
    setShowUpdateResultDialog: (state, action) => {
      state.showUpdateResultDialog = action.payload
    },
    resetState: () => initialState,
  }
})

export const {
  resetState,
  setShowUpdateResultDialog
} = dialogSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectShowUpdateResultDialog = state => {
  return state.dialog.showUpdateResultDialog
}

export default dialogSlice.reducer
