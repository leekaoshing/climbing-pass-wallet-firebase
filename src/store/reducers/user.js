import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  passes: null,
  updateResult: null
}

export const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    addGymToUser: (state, action) => {
      state.passes[action.payload] = 0
    },
    setPassesForGym: (state, action) => {
      const gymId = action.payload.gymId
      const numberOfPasses = action.payload.numberOfPasses
      state.passes[gymId] = numberOfPasses
    },
    setPassesForUser: (state, action) => {
      state.passes = action.payload
    },
    setUpdateResult: (state, action) => {
      state.updateResult = action.payload
    },
    resetState: () => initialState,
  }
})

export const {
  addGymToUser,
  resetState,
  setPassesForGym,
  setPassesForUser,
  setUpdateResult
} = userSlice.actions

// The function below is called a selector and allows us to select a value from
// the state. Selectors can also be defined inline where they're used instead of
// in the slice file. For example: `useSelector((state) => state.counter.value)`

export const selectUserPasses = state => {
  return state.user.passes
}

export const selectUpdateResult = state => {
  return state.user.updateResult
}

export default userSlice.reducer
