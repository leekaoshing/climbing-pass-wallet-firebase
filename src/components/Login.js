import { CircularProgress } from '@material-ui/core';
import Button from '@material-ui/core/Button';
import Paper from '@material-ui/core/Paper';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Alert from '@material-ui/lab/Alert';
import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { useFirebase, useFirestore } from 'react-redux-firebase';
import { setShowPasswordResetDialog } from "../reducers/dialogSlice";
import { setEditableUser } from "../reducers/userSlice";
import { PasswordResetDialog } from './PasswordResetDialog';
import { SignUpButton } from './SignUpButton';

const useStyles = makeStyles((theme) => ({
    paper: {
        padding: theme.spacing(2),
        textAlign: 'center',
    },
    submitPaper: {
        padding: theme.spacing(0),
        textAlign: 'center',
    },
}));

export function Login() {
    const dispatch = useDispatch();
    const classes = useStyles();
    const firebase = useFirebase();
    const firestore = useFirestore();

    const [identifier, setIdentifier] = useState('');
    const [password, setPassword] = useState('');
    const [identifierError, setIdentifierError] = useState('');
    const [passwordError, setPasswordError] = useState('');

    const [loginError, setLoginError] = useState('');
    const [loading, setLoading] = useState(false);

    const cannotSubmit = identifierError !== '' || passwordError !== '' || identifier === '' || password === '';

    const signInWithEmailAndPasswordHandler = async (email, password) => {
        console.log('USE FIREBASE', firebase);
        const auth = await firebase.auth().signInWithEmailAndPassword(email, password);
        const user = await firestore.collection('users').doc(auth.uid).get(); // TODO This is for login, but shouldn't be here since user may already be logged in on page load
        console.log('FETCHED USER', user.data())
        dispatch(setEditableUser(user.data()));
        // .then(() => {
        //     setIdentifier('');
        //     setPassword('');
        //     setLoading(false);
        // })
        // .catch(error => {
        //     setLoginError(error.message);
        //     setLoading(false);
        // });
    };

    const handleSubmit = () => {
        // TODO Cannot sign in by displayName yet, see getEmailFromDisplayName() in firebase.js
        // if (identifier.match(/^\S+@\S+[\.][0-9a-z]+$/g)) {
        //     auth.signInWithEmailAndPassword(identifier, password).catch(error => {
        //         setError("Error signing in with password and email!");
        //         console.error("Error signing in with password and email", error);
        //     });
        // } else {
        //     email = getEmailFromDisplayName(identifier);
        //     auth.signInWithEmailAndPassword(displayName, password).catch(error => {
        //         setError("Error signing in with password and email!");
        //         console.error("Error signing in with password and email", error);
        //     });
        // }

        // TODO add loading icon and error message
        setLoading(true);
        signInWithEmailAndPasswordHandler(identifier, password)
            .then(() => {
                setIdentifier('');
                setPassword('');
                setLoading(false);
            })
            .catch(error => {
                console.log('ERROR', error)
                setLoginError(error.message);
                setLoading(false);
            });
    }

    const handleKeyDown = event => {
        if (event.key === 'Enter' && !cannotSubmit) {
            handleSubmit();
        }
    }

    const handleForgotPassword = () => {
        dispatch(setShowPasswordResetDialog(true));
    }

    const onChangeHandler = event => {
        const { name, value } = event.target;
        if (name === "identifier") {
            setIdentifier(value);
        } else if (name === "password") {
            setPassword(value);
        }
    };

    const validateIdentifierOnBlur = () => {
        if (identifier === '') {
            setIdentifierError("Email cannot be blank");
        } else {
            setIdentifierError('');
        }
    }

    const validatePasswordOnBlur = () => {
        if (password === '') {
            setPasswordError("Password cannot be blank");
        } else {
            setPasswordError('');
        }
    }

    return (
        <div>
            <PasswordResetDialog />
            {
                loginError !== '' ?
                    <Paper className={classes.paper} elevation={0}>
                        <Alert severity="error">{loginError}</Alert>
                    </Paper>
                    :
                    null
            }
            <Paper className={classes.paper} elevation={0}>
                <TextField
                    error={identifierError !== ''}
                    label="Email"
                    id="email for login"
                    name="identifier"
                    helperText={identifierError}
                    variant="outlined"
                    value={identifier}
                    onChange={onChangeHandler}
                    onBlur={validateIdentifierOnBlur}
                    onKeyDown={handleKeyDown}
                />
                <br />
                <br />
                <TextField
                    error={passwordError !== ''}
                    label="Password"
                    id="password for login"
                    name="password"
                    helperText={passwordError}
                    variant="outlined"
                    type="password"
                    value={password}
                    onChange={onChangeHandler}
                    onBlur={validatePasswordOnBlur}
                    onKeyDown={handleKeyDown}
                />
            </Paper>
            <Paper className={classes.submitPaper} elevation={0}>
                {
                    loading ?
                        <CircularProgress />
                        :
                        <div>
                            <Button
                                disabled={cannotSubmit}
                                aria-label="submitLogIn"
                                variant="contained"
                                color="primary"
                                onClick={handleSubmit}
                            >
                                Log In
                            </Button>
                            <br />
                            <br />
                            <SignUpButton />
                            &nbsp;
                            &nbsp;
                            <Button
                                aria-label="forgotPassword"
                                variant="contained"
                                onClick={handleForgotPassword}
                            >
                                Forgot Password?
                            </Button>
                        </div>
                }
            </Paper>
        </div>
    );
};