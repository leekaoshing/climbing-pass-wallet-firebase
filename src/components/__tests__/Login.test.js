import {
    fireEvent, render,
    screen
} from '@testing-library/react';
import { setShowAddGymDialog } from '../../reducers/userSlice';
import wrapComponent, { makeStore } from '../../testUtils/testUtils';
import { Login } from '../Login';

// const mockFirebase = {
//     auth: jest.fn().mockReturnThis(),
//     signInWithEmailAndPassword: jest.fn((email, password) => { uid: 'someUid' })
// };

// const mockFirestore = {
//     collection: jest.fn().mockReturnThis(),
//     doc: jest.fn().mockReturnThis(),
//     get: jest.fn().mockReturnThis(),
//     data: jest.fn(() => mockuser)
// }

jest.mock('react-redux-firebase', () => ({
    useFirebase: {
        auth: jest.fn().mockReturnThis(),
        signInWithEmailAndPassword: jest.fn((email, password) => { uid: 'someUid' })
    },
    useFirestore: {
        collection: jest.fn().mockReturnThis(),
        doc: jest.fn().mockReturnThis(),
        get: jest.fn().mockReturnThis(),
        data: jest.fn(() => mockuser)
    }
}));


describe('Login', () => {


    it('renders and performs login action correctly', () => {
        const mockUser = {
            firstName: 'Adam',
            lastName: 'Janson',
            displayName: 'ajan',
            passes: {
                'BM': 1
            }
        }

        const initialState = {
            user: {
                editableUser: {
                    passes: {}
                },
            },
        };

        const store = makeStore(initialState);
        const component = wrapComponent(Login, store);



        const { asFragment, rerender } = render(component);
        expect(asFragment()).toMatchSnapshot();

        fireEvent.change(screen.getByLabelText('Email'), { target: { value: 'ajan@mail.com' } });
        fireEvent.change(screen.getByLabelText('Password'), { target: { value: '123456' } });
        fireEvent.click(screen.getByText(/Log In/i));

        // expect(mockFirebase.auth).toHaveBeenCalled();

        expect(store.getState().user.editableUser).toStrictEqual({
            passes: {
                'FB': 0
            }
        });
    });
})