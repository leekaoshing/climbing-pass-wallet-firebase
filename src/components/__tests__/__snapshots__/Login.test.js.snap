// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders and performs login action correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiPaper-root makeStyles-paper-1 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="email for login"
          id="email for login-label"
        >
          Email
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="email for login"
            name="identifier"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <br />
      <br />
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="password for login"
          id="password for login-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="password for login"
            name="password"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root makeStyles-submitPaper-2 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div>
        <button
          aria-label="submitLogIn"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Log In
          </span>
        </button>
        <br />
        <br />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Register
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
           
        <button
          aria-label="forgotPassword"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Forgot Password?
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Login renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiPaper-root makeStyles-paper-1 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="email for login"
          id="email for login-label"
        >
          Email
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="email for login"
            name="identifier"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Email
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <br />
      <br />
      <div
        class="MuiFormControl-root MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          for="password for login"
          id="password for login-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input"
            id="password for login"
            name="password"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-7"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiPaper-root makeStyles-submitPaper-2 MuiPaper-elevation0 MuiPaper-rounded"
    >
      <div>
        <button
          aria-label="submitLogIn"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Log In
          </span>
        </button>
        <br />
        <br />
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Register
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
           
        <button
          aria-label="forgotPassword"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Forgot Password?
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
